### ISSUE TITLE
In 1000 node clusters some nodes fail to register/become running

### ISSUE THREAD

**Issue by gmarek** [kubernetes/kubernetes#18969]
http://kubekins.dls.corp.google.com/job/kubernetes-e2e-gce-enormous-cluster/7/consoleFull

This is a blocker for a 1000-node effort.

cc @kubernetes/goog-control-plane @kubernetes/goog-node @brendandburns @davidopp @wojtek-t @fgrzadkowski

**Comment by gmarek** [comment]
It turns out that it happens even on a 100 node cluster:
http://kubekins.dls.corp.google.com/view/Critical%20Builds/job/kubernetes-e2e-gce-scalability/3725/
http://kubekins.dls.corp.google.com/view/Critical%20Builds/job/kubernetes-e2e-gce-scalability/3724/
http://kubekins.dls.corp.google.com/view/Critical%20Builds/job/kubernetes-e2e-gce-scalability/3721

It started to happen after increasing verbosity level in kubelet to `--v=4`, but I can't see how that may be related, as in 1000-node cluster we're running with `--v=1`, and increasing kubelet verbosity doesn't impact API server in any way (I hope). cc @wojtek-t 

@dchen1107 - please fix or reasign.

**Comment by wojtek-t** [comment]
My feeling is that it's not strictly related to logs verbosity - it seems that sometimes some Kubelets fail to register themself. It's obviously more probably in scalability tests, because there are way more Kubelets there, so bigger change for the failure...

**Comment by gmarek** [comment]
Another failure: http://kubekins.dls.corp.google.com/job/kubernetes-e2e-gce-enormous-cluster/23

**Comment by timothysc** [comment]
@gmarek I can not access that URL listed above.  

/cc @kubernetes/sig-scalability

**Comment by quinton-hoole** [comment]
@timothysc the logs are at:

https://cloud.google.com/console/storage/kubernetes-jenkins/logs/kubernetes-e2e-gce-enormous-cluster/23/

**Comment by gmarek** [comment]
Main problem is that there are no logs...

**Comment by dchen1107** [comment]
I am shocked the node registration has the dependency with kubelet's log level. But under e2e circumstances, it might be possible due to slow running kubelet with enormous logging, and reaches some timeout values here. In the plane, cannot check our Jenkins builds yet.

**Comment by gmarek** [comment]
@quinton-hoole said that we probably shouldn't worry about this failures, as they are rare in GCE. If we fail in one of 10 runs of 1000 machines and only 1 or two of them are not responsive it means that we succeed 99.98% of times - which is most likely good enough.

OTOH Samsung said they have much higher failure rate, which may mean that there's some underlying issue: @spiffxp @countspongebob

**Comment by dchen1107** [comment]
Checked kubernetes-e2e-gce-enormous-cluster jenkin project, and there is only one failed case posted by @gmarek above. Also only node in that failed case report NodeNotReady:

```
e2e-enormous-cluster-minion-btjs   failure-domain.alpha.kubernetes.io/region=asia-east1,failure-domain.alpha.kubernetes.io/zone=asia-east1-a,kubernetes.io/hostname=e2e-enormous-cluster-minion-btjs   NotReady   8m
```

I am closing this issue for now. Please feel free to re-open it if the issue recurred.

**Comment by gmarek** [comment]
It happened again: http://kubekins.dls.corp.google.com/view/Scalability/job/kubernetes-e2e-gce-enormous-cluster/30/

Reason is exactly the same, namely one node is NotReady:
```
21:00:43 e2e-enormous-cluster-minion-d24g   failure-domain.alpha.kubernetes.io/region=asia-east1,failure-domain.alpha.kubernetes.io/zone=asia-east1-a,kubernetes.io/hostname=e2e-enormous-cluster-minion-d24g   Ready      37m
21:00:43 e2e-enormous-cluster-minion-d3oh   failure-domain.alpha.kubernetes.io/region=asia-east1,failure-domain.alpha.kubernetes.io/zone=asia-east1-a,kubernetes.io/hostname=e2e-enormous-cluster-minion-d3oh   NotReady   9m
```
Note the age difference.

**Comment by quinton-hoole** [comment]
I would speculate that there's an intermittent problem with that one VM
(e.g. a faulty NIC, or otherwise dodgy host machine.  One intermittently
dodgy VM out of several thousands is not, IMO, something we should be
worrying about.

**Comment by wojtek-t** [comment]
Yes and no. If even if 1 VM will not start correctly, we claim that the whole cluster didn't started correctly and finish the startup with failure, this isn't ok in my opinion. Maybe it would be ok to finish with success with warnings in that case, but we should do something...

**Comment by quinton-hoole** [comment]
@wojtek-t Yes, agreed, that's what I was trying to say.  My apologies for being unclear.  A fraction of a percentage of VM's not coming up healthy should be considered normal, and not cause the failure of the test.

**Comment by davidopp** [comment]
Why was this issue closed? I don't have a problem with changing the test to not fail, but shouldn't we try to root cause the underlying problem of some nodes not registering? I don't disagree with Quinton's speculation, but it *is* speculation...

**Comment by quinton-hoole** [comment]
@davidopp My opinion is that:

1. Irrespective of what the root cause is in these particular cases (there may be one root cause, or several), we still need to update the test to not fail if an acceptably small percentage (0.1% ?) of nodes do not come up healthy.
2. Even if we do root cause and fix this particular problem, we will see other intermittent VM problems in future.  Do we try to root cause every time we see a node not come up healthy?
3. As long as the incidence of unhealthy nodes is in the 0.01% range, the time and effort required to root cause the problem each time would be much better spent on other things.

So in summary, I'd suggest that once 1 above is in place, we should close this ticket.

**Comment by gmarek** [comment]
I't happening WAY more frequently now. Pretty much every second run, which means we're on 99,95%, which, I guess is a decent SLO, with which we can leave with assuming that scripts will be more resilient to failures. Opened #19786 for this.

**Comment by brendandburns** [comment]
This happened again:

**Comment by brendandburns** [comment]
We should at least try to grab kubelet and dmesg of the nodes that are failing...

**Comment by bprashanth** [comment]
and startup.log if possible, in case GCS or some other storage location a salt script tries to access 500s

**Comment by dchen1107** [comment]
Ok, did a little bit data processing for recent failure mentioned by @brendandburns above:

Sort and count elements (hostnames) after merging two strings in build.log: "01:27:37 NODE_NAMES=..." and output of kubectl get nodes, found 
...
      2 e2e-scalability-minion-w1hw
      2 e2e-scalability-minion-w1st
      2 e2e-scalability-minion-w3hp
      1 e2e-scalability-minion-w5r2                    <=======
      2 e2e-scalability-minion-wjhp
      2 e2e-scalability-minion-wjke
...

There is no NodeSpec object created for e2e-scalability-minion-w5r2. This case somewhat proves @quinton-hoole's theory above correct: VM itself doesn't come up here. 

I can try to add the logic to grap kubelet or dmesa of that bad node, but don't think it will always work. Also in most of time, processing those data is wasting our time here.

**Comment by fgrzadkowski** [comment]
Isn't it that we wait for MIG to actually start all VMs before we do any kube-level verification?

https://github.com/kubernetes/kubernetes/blob/master/cluster/gce/util.sh#L739

**Comment by dchen1107** [comment]
The initial version I created for kube-level verification did make sure the node is up. But after switching MIG, and introducing config-vm.sh, I don't think we did that. At least I didn't see it from our scripts. Maybe MIG will report failure itself, and I am not aware it?

**Comment by dchen1107** [comment]
We failed at https://github.com/kubernetes/kubernetes/blob/master/cluster/validate-cluster.sh#L62
But this script is used by production and other cloud providers. I definitely don't want to mess up validate_cluster.sh here. 

Also there is no proper exit code indicate the detail failure reason. On script exit with code 2 could mean a lot of other reasons.

**Comment by davidopp** [comment]
@dchen1107 I didn't fully understand your response to Filip's question. The line he pointed to calls `gcloud compute instance-groups managed wait-until-stable` from `create-nodes()`. kube-up.sh doesn't do validate-cluster until after that. So how can we be calling validate-cluster before all 1000 nodes are up? Do you think one came up but then already failed before we validate cluster?

**Comment by dchen1107** [comment]
The point I had is that I am not sure about MIG signal on stable. Is that node up and finish run config-vm.sh, which does a lot of stuff on the node, including installing kubelet binary? Before we are switching to MIG, we checked the node up first, then check Kubelet and docker up or not. Now all logic are removed from validate_cluster.sh since we handle them separately. This is good part. But on another side, we still use validate_cluster to check if Kubelet is ready here or not, if no, bail out earlier. In this case, we hide a lot of details.

**Comment by davidopp** [comment]
Thanks for explaining. (BTW I assume "config-vm.sh" is referringn to "cluster/gce/configure-vm.sh" ?

Should we make config-vm.sh not exit until node, kubelet, and docker are confirmed healthy? It seems that if we could get that into a log somewhere, then we could at least know which component is not coming up, and also if those all pass, then we know the problem is somewhere later.

**Comment by dchen1107** [comment]
> (BTW I assume "config-vm.sh" is referringn to "cluster/gce/configure-vm.sh" ?

yes.

>Should we make config-vm.sh not exit until node, kubelet, and docker are confirmed healthy? It seems that if we could get that into a log somewhere, then we could at least know which component is not coming up, and also if those all pass, then we know the problem is somewhere later.

I am not sure I understand what you suggested above. validate-cluster.sh is invoked by kube-up.sh through e2e.go. config-vm.sh is running at each node. The issue here is that validate-cluster.sh exit with code 2, which could be many reasons: VM failed to up, kubelet install issue, podcidr allocation issue, docker is done, ...

**Comment by davidopp** [comment]
I guess there are two things I am suggesting:
(1) Can we have config-vm.sh do more thorough health-checking of the node components (docker, kubelet, etc.) and logging of the result, so that we can easily find out if there was a problem with docker or kubelet or no problem, without having to dig through their logs
(2) Can we have whatever signal `gcloud compute instance-groups managed wait-until-stable` uses from each node, not return good until config-vm.sh has done health-checking of the node components. 

Sorry if these are dumb questions; I am not very familiar with this code.

**Comment by roberthbailey** [comment]
For (1), right now the last thing that `configure-vm.sh` does is run salt. Then it exits. I don't think it would be too difficult to add a function to hit the kubelet's `/healthz` endpoint and print out a message once it thinks the kubelet is happy. I'm not sure if the kubelet API exposes enough information for the script to block until the kubelet has successfully registered itself with the master, reconfigured `cbr0`, restarted docker, and become ready. 

(2) depends on functionality of MIGs. From their [docs](https://cloud.google.com/compute/docs/instance-groups/) I don't see a way to configure any sort of health check other than that the VM is running.

**Comment by fgrzadkowski** [comment]
I've just discussed this with Managed Infra folks in WAW and currently there is no generic mechanism to wait for startup script to complete before marking VM as running. In other words it's possible that we might be verifying cluster without waiting for kubelets to even start. The only strange thing is that after all VMs are running (i.e. OS is started) we still wait couple of minutes for initialization to happen. Why does it take so long for `configure-vm.sh` to run?

**Comment by quinton-hoole** [comment]
@fgrzadkowski @davidopp @roberthbailey @dchen1107 
I just took a quick look at kubernetes-e2e-gce-scalability, and in all the history we have (132 runs at 100 nodes per run = 13,200 nodes) there is one single node failure like this (the one @brendandburns reported above).  That's a 0.007% failure rate.  As I mentioned [above](https://github.com/kubernetes/kubernetes/issues/18969#issuecomment-172257523), I would strongly suggest that we amend the test to tolerate single-node failures, and spend our time on more important things.

Am I missing something?

**Comment by davidopp** [comment]
Yes, I think a slightly modified version of what @quinton-hoole suggested is good.

In particular, I think we should create two versions of this test, one that tolerates single-node failures and one that doesn't. Put the first one in the critical suite and block merges on it, and put the second one in the flaky suite and mark it a P1 and assign someone on node team or control-plane team to make it work 100% ASAP. There are some ideas earlier in this thread on how to do it, but let's not debate the exact approach now. Once they are able to do that, we can make the critical/blocking version of the test expect 100% of nodes, and get rid of the flaky one.

The end-user experience goal *has* to be all 1000 nodes come up 100% of the time, even if it requires retries at multiple levels of the stack and has a very long 99%ile turnup time due to the retries. We can iterate on that over time to make it better (e.g. initially we can just kube-down the whole cluster if all 1000 nodes don't come up, and kube-up again, and repeat until 1000 nodes come up; retries lower in the stack can be thought of as optimizations to make the process faster and lightweight).

@dchen1107 @quinton-hoole does that sound reasonable?

**Comment by davidopp** [comment]
(And @dchen1107 had also suggested a more tolerant version of this test at the war room this morning.)

**Comment by dchen1107** [comment]
@quinton-hoole I agreed with you completely. This is also the point I tried to share with everyone in today's warroom meeting. 

Another if we look at history more closely, instance-group failed more often than one node not coming up after instance-group is stabled.

**Comment by quinton-hoole** [comment]
@davidopp I don't think that the effort and resource expense required to achieve 100% of 1,000 node clusters coming up with all nodes healthy is justified.  I think that we've already spend far too much time on this issue.  I would not suggest spending any more than that required to make the test tolerant of single-node failures, as described above.

**Comment by davidopp** [comment]
I disagree. Would you as a user consider it acceptable if asking GCE to bring up 1000 VMs sometimes only brought up 999? I wouldn't be, and I feel the same way about Kubernetes nodes.

I totally sympathize with the test issue, and I'm all for allowing a node failure in the "main" version of the test. But I think it's worth debugging why we can't reliably bring up 1000-node clusters, and having an intolerant version of the test in the flaky suite will let us do that. Don't worry, you don't have to be the one to work on it. :)

**Comment by dchen1107** [comment]
@davidopp Can we have SLO with more detail here? 

Even we said having all 1000 nodes come up 100% of the time as the goal for GKE. There are many other ways to do it. For example, pushing harder to make instance group updatable, or after a while, and start another instance group with number of missing nodes? Shouldn't using node auto register is relatively easy to solve the problem here? And there is different from the end user's perspective. 

On another side, I think the cluster up time should take into consideration too. Repeating kube-down and kube-up might violate this goal (if we consider this as goal).  Also the cost of iterating of kube-down & kube-up might be much higher than what you thought here. Thinking about API quota, and other resource limits?

Back to the test itself, from above @fgrzadkowski and @roberthbailey's reply, you can see that they approve my theory that config-vm.sh itself might not be finished. Actually it is very easy to reproduce by spining your own cluster. Another point no one brought up is that even instance group said it is stable, it only means it receives one signal from each node, which doesn't include vm crash right after this. We just recent saw this issue, right? 

Also people above mentioned that collecting config-vm logs. It is available at /var/log/startupscript.log including all salt related information. The problem is that it might only be available through node's serial console when vm is crash. @roberthbailey @ihmccreery on this case, can collect serial console log on this case?

**Comment by roberthbailey** [comment]
While I agree that configure-vm not finishing might be the *cause* of all nodes not being healthy and we should make it easier to debug, I agree with @davidopp that we shouldn't tolerate a user asking for N nodes and only providing them with X% (where X<100) of them and claiming success. We certainly wouldn't do this for GKE; why would we do it for OSS?

**Comment by dchen1107** [comment]
Sending above response before see both @quinton-hoole and @davidopp's last response:

https://github.com/kubernetes/kubernetes/issues/18969#issuecomment-176952012 &
https://github.com/kubernetes/kubernetes/issues/18969#issuecomment-176955041

I agreed with @quinton-hoole here, it is not worthy the engineering effort on debugging one node failure when the failure rate is < 0.1% here. We should treat this kind of test as **performance regression test**, which can define some small percentage on failure rate. 

This one test effectively exercises many components in the stack:
1) managed-instance-group correctness and performance
2) containervm image
3) config-vm.sh including salt
4) Kubelet auto-join feature
5) master component's podCIDR's allocation
6) healthiness of docker as container runtime 
7) kube-daemons auto recovery through supervisord
...

Except 1) out of our control, should every single bullet listed we have specific test covered? For example, every single e2e build examine kubelet's node auto-join path, including reboot which checks rejoin here. Same for podCIDR's allocation. Shouldn't we detect one logic failure lately through a nodecontroller specific test, not this one?

Since we have that many more specific tests written to cover our system, why not treat this as performance regression test since we did observe that all components might fail due to the reasons out of our control, for example, network partition, kernel coredump, etc.

If we absolutely hate the idea of performance regression test here. Then let's add failure handling code for this test, something like:
1) ask instance-groups about N nodes
2) gcloud compute instance-groups managed wait-until-stable
3) check how many nodes in ready state periodically until reaching the timeout value (T1)
4) once hitting T1, find the number (m) of the outliers, and resize your cluster to N+m. Meanwhile, logging the information of those outliers. Not sure if we can remove those bad sheep from instance-group or not. 
...

This is a large distributed system, we cannot assume there is no failure in the entire stack. Instead, we should detect those failure and handle them. Current test is written in a way of everything is perfect, which is wrong.

**Comment by davidopp** [comment]
I would suggest that we take this discussion offline, because we're not coming to agreement here.

I've asked @gmarek to clone the test, modify the current version so that it accepts a small number of nodes not coming up (as requested by @dchen1107 and @quinton-hoole), and put the version that requires all nodes to come up into the flaky suite. This will satisfy the requirements of the flake fixit. Offline we can discuss the longer-term plan.

**Comment by bgrant0607** [comment]
Failing the whole test is a poor signal. Reporting the number of node failures would provide more information, and we could then set a realistic threshold to guard against regressions, which we could improve over time.

Yes, we should also do what we can to make the failures easier to debug, including making more specific tests, more logging, allowing more failures and aggregating results, etc.

We should also add self-healing, so that nodes that fail to come up are replaced/repaired, generally, not in the test.

We should also file GCE bugs for any problems traced to the underlying infrastructure -- cc: me on the issues.

**Comment by ihmccreery** [comment]
> Failing the whole test is a poor signal. Reporting the number of node failures would provide more information, and we could then set a realistic threshold to guard against regressions, which we could improve over time.

For what it's worth, @smarterclayton has done something very similar for large-scale namespace deletion.

**Comment by bgrant0607** [comment]
I'm also 100% certain users would much rather have 999 nodes come up and to have a report saying one failed, with that node being replaced and coming up later, than to have the whole cluster creation process abort in the middle.

**Comment by gmarek** [comment]
I'm not really sure what are we discussing here. IIUC all comments we all agree that we need to make sure that we start up 100% of requested nodes, but we're OK with tolerating small number of failed nodes in tests until we reach this point. 

There're a lot of ways how to get to 100% so maybe we should assign this to a single person and leave the decision on how to proceed up to him/her?

**Comment by dchen1107** [comment]
@gmarek Thanks for #20407. But this is different from what I proposed above. If we want to treat the test regarding node up as performance regression test, and tolerate small number of failed nodes in node, like one out of 10,000 nodes, we should handle it at higher level, like test harness itself, instead of validate_cluster.sh here. Thinking about our scalability test, we only bring up 100 nodes for each run, how am I going to config ALLOWED_NOTREADY_NODES for each run? 

I am going to file a separate issue on error handling for kube-up.

**Comment by gmarek** [comment]
@dchen1107 - I'm not really sure what's your question. We set environmental variables in each suite and we'll continue to do so. Whatever we do on the higher level we want to deflake tests, which requires changes in validate-cluster.sh.

I can easily enough write higher-level regression tests that will parse test logs, but we'd need to decide on what exactly our goal is.

**Comment by davidopp** [comment]
Please please please please please please please please please please please let's stop discussing this on Github. I will schedule an in-person meeting.

**Comment by fgrzadkowski** [comment]
Changing priority to P1 as we have a short term solution (#20407) which allows some nodes to fail.

**Comment by dchen1107** [comment]
This is not flake, instead of a broken feature or bad test like I mentioned above. We should fix the feature by introducing error handling here. Remove flake label.

**Comment by gmarek** [comment]
As for how to proceed: the question left is if we want to enable this solution for production clusters (i.e. set defaults to non-0).

We have `enormous-cluster-startup` test running daily (for first few days it was failing, because it tried to find JUnit results - my fault). Afterwards it was running, until first failure (one node didn't get registered). Sadly after that it wasn't able to start, as teardown wasn't able to clean up all resources:
```
08:35:36  - Instance 'e2e-enormous-startup-minion-uxql' deletion failed: The resource 'projects/kubernetes-scale/zones/us-east1-b/instances/e2e-enormous-startup-minion-uxql' is not ready
```

cc @kubernetes/goog-control-plane

**Comment by dchen1107** [comment]
We can talk about more on the default config for production cluster. But even without changing the default value, we should document this: #21000 or through other document, so that the user can config this.

**Comment by dchen1107** [comment]
cc/ @roberthbailey and @alex-mohr on the max number of node for a cluster on GKE side today? We can talk about the default config based on that max number.

On k8s side, I think document should cover it.

**Comment by davidopp** [comment]
Doesn't this already work in production on GCE, in that if you set `ALLOWED_NOTREADY_NODES` to some value (greater than zero) you get the behavior where kube-up will allow some failed nodes?

I think we should keep default as requiring all nodes to come up, and add documentation of `ALLOWED_NOTREADY_NODES`.

**Comment by dchen1107** [comment]
@davidopp yes, we are on the same page for k8s. @gmarek is asking if we should provide a default non-0 for production.

**Comment by davidopp** [comment]
Oh I didn't realize "production" == "GKE" :)

I don't think it should ever default to non-zero. But I'm not sure how much work it would be to allow people to specify the value of `ALLOWED_NOTREADY_NODES` at cluster creation time on GKE (e.g. through the UI).

**Comment by roberthbailey** [comment]
Max # of nodes in GKE today is 250 (which was the tested limit for 1.1). We hope to increase that with 1.2 since k8s will support 1000.

**Comment by davidopp** [comment]
Until we have the background-join feature that @bgrant0607 has described elsewhere, I don't think we want cluster creation to succeed with less than the requested number of nodes.

Background-join should probably be a P1 issue to be done immediately post-1.2 though...

**Comment by dchen1107** [comment]
Here is the issue capturing your background-join: https://github.com/kubernetes/kubernetes/issues/20424

Maybe @bgrant0607 filed a different issue though?

**Comment by bgrant0607** [comment]
@davidopp @dchen1107 I haven't not filed another issue.

Kubelets already self-register. Broken nodes just need to be "healed" (rebooted or re-created) so that they can try to join again. And the turnup process needs to be tolerant of nodes not coming up immediately.

I agree that for now we should just document the existence of ALLOWED_NOTREADY_NODES.

That said:

This boils down to the common issue of fast failure vs. continuing on errors.

My opinion is that kube-up should continue unless no nodes come up, and should report how many nodes came up at the end. It could also give a non-zero exit code if any nodes don't come up.

How long does it take for a 1000-node cluster to come up? I assume it takes a while.

If the cluster is usable, I'd prefer to be able to use the nodes available, while investigating why other nodes didn't come up, or simply retry those nodes.

If 997 out of 1000 nodes come up and the whole cluster control plane comes up, don't error out in the middle and leave me wondering what to do next. Do I need to tear it down and start over? Do I need to manually clean up? Can I retry without tearing it down? Can I use what is there? 

What if the user is bringing up a cluster on bare metal, and one of the nodes is broken?

What if someone is trying to bring up a cluster on spot instances and one VM gets killed?

What if Docker wedges on a few nodes?

It's even more important to be tolerant of failures for real user clusters than in tests.

**Comment by davidopp** [comment]
IIUC we're not planning to do anything more on this for v1.2, so I'm moving it to next-candidate.

**Comment by davidopp** [comment]
Oh, actually I guess there is one remaining v1.2 question, namely the one @gmarek asked (and @dchen1107 repeated), whether given that we're going to 1k nodes in GKE do we want some way for the user to say "it's OK if I don't get 1000 nodes" when they spin up a GKE cluster.

**Comment by gmarek** [comment]
@alex-mohr @aronchick

**Comment by roberthbailey** [comment]
I think what we decide to do for GKE is beside the point for this issue (and we can track it internally).

**Comment by dchen1107** [comment]
Great, looks like we are converging on this for 1.2 release at least. I an punting this to next release.

**Comment by roberthbailey** [comment]
We need to track the documentation updates for the 1.2 release. Is there a separate issue for that?

**Comment by davidopp** [comment]
We have a special page on running large clusters
https://github.com/kubernetes/kubernetes/blob/master/docs/admin/cluster-large.md
so we can put anything relevant there.

I can send out a PR right now adding mention of `ALLOWED_NOTREADY_NODES` and fixing the mention of 250 node max to get the ball rolling (it pains me to file more issues given how many we already have) and we can discuss what else to add in that PR.

**Comment by roberthbailey** [comment]
As long as we have an issue or a PR I think we won't forget to finish it by 1.2.

**Comment by gmarek** [comment]
It happens even in our 100 Node clusters now:
http://kubekins.dls.corp.google.com/view/Scalability/job/kubernetes-e2e-gce-scalability/4823

**Comment by davidopp** [comment]
I'm moving this back to v1.2 and marking it P0 since it has started happening on 100 node clusters.

**Comment by gmarek** [comment]
It's getting worse...

https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/22047/kubernetes-pull-build-test-e2e-gce/30604/build-log.txt

**Comment by quinton-hoole** [comment]
Agreed - error rates this high go way beyond the 0.01% range we discussed before:

http://kubekins.dls.corp.google.com/view/Scalability/job/kubernetes-e2e-gce-scalability/4856/

```
02:03:31 Detected 96 ready nodes, found 100 nodes out of expected 100. Your cluster may not be fully functional.
```

In the above example the number of ready nodes climbed from 0 to 96 (of 100) in under 2 minutes, and then stuck at 96 for the rest of the 25 minute timeout.  So it's not a case of extending the timeout period, that's for sure.  In fact we should probably reduce it dramatically, from 25 minutes to about 3 minutes (for 100 node clusters at least - larger cluster might take a bit longer).  Something like 2 to 3 sec x number of nodes seems about right.

```
01:37:44 Waiting for 100 ready nodes. 0 ready nodes, 1 registered. Retrying.
01:37:59 Waiting for 100 ready nodes. 0 ready nodes, 31 registered. Retrying.
01:38:14 Waiting for 100 ready nodes. 0 ready nodes, 93 registered. Retrying.
01:38:30 Waiting for 100 ready nodes. 2 ready nodes, 100 registered. Retrying.
01:38:45 Waiting for 100 ready nodes. 44 ready nodes, 100 registered. Retrying.
01:39:00 Waiting for 100 ready nodes. 91 ready nodes, 100 registered. Retrying.
01:39:15 Waiting for 100 ready nodes. 96 ready nodes, 100 registered. Retrying.
...
02:02:45 Waiting for 100 ready nodes. 96 ready nodes, 100 registered. Retrying.
02:03:00 Waiting for 100 ready nodes. 96 ready nodes, 100 registered. Retrying.
02:03:16 Waiting for 100 ready nodes. 96 ready nodes, 100 registered. Retrying.
02:03:31 Detected 96 ready nodes, found 100 nodes out of expected 100. Your cluster may not be fully functional.

**Comment by ihmccreery** [comment]
@jlowdermilk looks like folks here had penned 0.01% as a reasonable threshold, just FYI.

**Comment by dchen1107** [comment]
I were out sick last two days, and agreed that failure rate is not acceptable. I am investigating it.

**Comment by dchen1107** [comment]
#18969  can really help with understanding the issue better.

**Comment by davidopp** [comment]
> #18969 can really help with understanding the issue better.

#18969 is this issue. Did you mean #10423?

**Comment by dchen1107** [comment]
@davidopp, Yes,  I mean #10423 or #21722. Sorry!

**Comment by dchen1107** [comment]
#22071 were closed with fix. I checked scalability tests since last Friday. It failed occasionally due to other reasons, but not #22071 anymore. Punt this to next release again.

**Comment by gmarek** [comment]
We failed cluster startup with 997 ready nodes: http://kubekins.dls.corp.google.com/view/Scalability/job/kubernetes-e2e-gce-enormous-cluster/76/

**Comment by gmarek** [comment]
We have a run into it in http://kubekins.dls.corp.google.com/view/Scalability/job/kubernetes-kubemark-gce-scale/830/console for 11 nodes. We do have logs from the kubelet that failed to become registered (jjgh), but they are empty. It looks like a VM issue, as kern.log is also empty. Or we somehow failed to download logs... @spxtr

**Comment by dchen1107** [comment]
cc/ @jlowdermilk For GKE node health tracking.
